gprofiler/log.py

```diff
@@ -0,0 +1,226 @@
+        self._path = path
+        self._logs: List[Dict] = []
+        self._logger = get_logger_adapter("gprofiler.RemoteLogsHandler")
+
+        # The formatter is needed to format tracebacks
+        self.setFormatter(logging.Formatter())
+
+    def init_api_client(self, api_client: 'APIClient'):
+        self._api_client = api_client
+
+    def emit(self, record: LogRecord) -> None:
+        if record.gprofiler_adapter_extra.pop(NO_SERVER_LOG_KEY, False):  # type: ignore
+            return
+
+        self._logs.append(self._make_dict_record(record))
+
+    def _make_dict_record(self, record: LogRecord):
+        formatted_timestamp = datetime.datetime.utcfromtimestamp(record.created).isoformat()
+        extra = record.gprofiler_adapter_extra  # type: ignore
+
+        # We don't want to serialize a JSON inside JSON but either don't want to fail record because of extra
+        # serialization, so we test if the extra can be serialized and have a fail-safe.
+        try:
+            json.dumps(extra)
+        except TypeError:
+            self._logger.exception(
+                f"Can't serialize extra (extra={extra!r}), sending empty extra", bad_extra=repr(extra)
+            )
+            extra = {}
+
+        run_id = extra.pop(RUN_ID_KEY, None)
+        if run_id is None:
+            self._logger.error("state.run_id is not defined! probably a bug!")
+            run_id = ''
+
+        cycle_id = extra.pop(CYCLE_ID_KEY, '')
+
+        assert self.formatter is not None
+        if record.exc_info:
+            # Use cached exc_text if available.
+            exception_traceback = (
+                record.exc_text if record.exc_text else self.formatter.formatException(record.exc_info)
+            )
+        else:
+            exception_traceback = ''
+
+        extra = extra if not extra.pop(NO_SERVER_EXTRA_KEY, False) else {}
+
+        return {
+            "message": record.message,
+            "level": record.levelname,
+            "timestamp": formatted_timestamp,
+            "extra": extra,
+            "logger_name": record.name,
+            "exception": exception_traceback,
+            RUN_ID_KEY: run_id,
+            CYCLE_ID_KEY: cycle_id,
+        }
+
+    def try_send_log_to_server(self):
+        assert self._api_client is not None, "APIClient is not initialized, can't send logs to server"
+        # Snapshot the current num logs because logs list might be extended meanwhile.
+        logs_count = len(self._logs)
+        try:
+            self._api_client.post(self._path, data=self._logs[:logs_count], api_version='v1')
+        except (APIError, RequestException):
+            self._logger.exception("Failed sending logs to server")
+        else:
+            # If succeeded, remove the sent logs from the list.
+            self._logs[:logs_count] = []
+
+
+class ExtraFormatter(logging.Formatter):
+    FILTERED_EXTRA_KEYS = [CYCLE_ID_KEY, RUN_ID_KEY, NO_SERVER_LOG_KEY, NO_SERVER_EXTRA_KEY]
```

Comment:
Any idea how these can be made static? so they're really part of the functions' signatures?

Reply:
Not really, the extra arguments are just a dict inside the `LogRecord` I can't use python keyword arguments binding for that :(